feat: use CoW for List and Table contents

cgzones opened the pull request:
Use a Copy-on-Write type for the contents of tables and lists so their content can be cached and only recomputed on actual changes.  This avoids unnecessary computations for the rendering in case of unchanged data.

Related to ratatui/ratatui-table#5.

Contains first commit of ratatui/ratatui#1791, with `'lend` and `'data` lifetimes combined into `'a`.

joshka reviewed (commented):
For the macros, it could be worth putting these in the ratatui-macros project which has some stuff for table rows already. WDYT?

https://docs.rs/ratatui-macros/latest/ratatui_macros/#row

I think in general, this idea seems fine. The main question I'd have are about how this affects existing code. I'm assuming that it should continue to work fine, but there's lots of changes to constructor type code which I'm unsure whether it was due to necessity or preference. Leaving a few clear tests that show the existing call patterns being undisturbed would generally allay that concern.

cgzones commented on ratatui-widgets/src/table.rs:
```diff
@@ -271,10 +272,44 @@ pub struct Table<'a> {
+        ([] as [$crate::table::Row<'_>; 0])
+    );
+    ($([$($x:expr),+ $(,)?]),+ $(,)?) => (
+        // TODO: avoid to_vec()
```
to avoid the allocation in `.to_vec()`

cgzones commented on ratatui-widgets/src/list.rs:
```diff
@@ -142,6 +145,41 @@ pub enum ListDirection {
+///     .repeat_highlight_symbol(true);
+///
+/// frame.render_widget(list, area);
+/// # }
```
Since `ListItem` does not implement copy, the trivial implementation does not work

cgzones commented on ratatui-widgets/src/list/list_items.rs:
When trying to extract the macro I get the following errors:
```
error[E0271]: type mismatch resolving `<[ListItem<'_>] as ToOwned>::Owned == Vec<ListItem<'_>>`
   --> ratatui-widgets/src/list.rs:743:40
    |
743 |         let cow: Cow<[ListItem<'_>]> = Cow::Owned(items);
    |                                        ^^^^^^^^^^^^^^^^^ expected `ratatui::widgets::ListItem<'_>`, found `item::ListItem<'_>`
    |
note: the crate `ratatui_widgets` is compiled multiple times, possibly with different configurations
   --> ratatui-widgets/src/list/item.rs:73:1
    |
 73 | pub struct ListItem<'a> {
    | ^^^^^^^^^^^^^^^^^^^^^^^ this is the found type `item::ListItem`
    |
   ::: /var/home/user/Coding/workspaces/ratatui/ratatui-widgets/src/list/item.rs:73:1
    |
 73 | pub struct ListItem<'a> {
    | ^^^^^^^^^^^^^^^^^^^^^^^ this is the expected type `ratatui::widgets::ListItem`
    |
   ::: ratatui-widgets/src/list.rs:523:9
    |
523 |     use ratatui::macros::list_items;
    |         ------- one version of crate `ratatui_widgets` used here, as a dependency of crate `ratatui_macros`
    = help: you can use `cargo tree` to explore your dependency tree
```

orhun replied:
That looks odd. Are you sure you are using `ratatui-widgets` from the workspace?

cgzones committed:
feat(list): use CoW for content

Use a Copy-on-Write type for the contents of lists so their content can
be cached and only recomputed on actual changes.
This avoids unnecessary computations for the rendering in case of
unchanged data.

Related to #1004.

cgzones committed:
feat(table): use CoW for content

Use a Copy-on-Write type for the contents of tables so their content can
be cached and only recomputed on actual changes.
This avoids unnecessary computations for the rendering in case of
unchanged data.

Related to #1004.

cgzones committed:
feat(list): add list_items! macro

cgzones committed:
feat(row): use CoW for content

Use a Copy-on-Write type for the contents of table rows so their
content can be cached and only recomputed on actual changes.
This avoids unnecessary computations for the rendering in case of
unchanged data.

Related to #1004.

cgzones committed:
feat(row): add row_cells! macro

cgzones committed:
feat(table): add table_rows! macro

cgzones committed:
feat(examples): make use of CoW lists and tables
